File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
int height(TreeNode* root,int &diameter){
    if(root==NULL){
        return 0;
    }
    int Hleft=height(root->left,diameter);
    int Hright=height(root->right,diameter);

    diameter=max(diameter,Hleft+Hright);
    return 1+max(Hleft,Hright);

}
    int diameterOfBinaryTree(TreeNode* root) {
        int diameter=0;

        height(root,diameter);
        return diameter;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    static int height(int[] diameter,TreeNode root){
        if(root==null){
            return 0;
        }
        int lh=height(diameter,root.left);
        int rh=height(diameter,root.right);
        diameter[0]=Math.max(diameter[0],lh+rh);
        return 1+Math.max(lh,rh);
    }
    public int diameterOfBinaryTree(TreeNode root) {
        int[] diameter =new int[1];
        height(diameter,root);
        return diameter[0];
        
    }
}